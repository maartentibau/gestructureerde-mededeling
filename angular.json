{
  "$schema": "./node_modules/@angular/cli/lib/config/schema.json",
  "version": 1,
  "newProjectRoot": "projects",
  "projects": {
    "gestructureerde-mededeling": {
      "projectType": "application",
      "schematics": {
        "@schematics/angular:component": {
          "style": "scss"
        }
      },
      "root": "",
      "sourceRoot": "src",
      "prefix": "app",
      "architect": {
        "build": {
          "builder": "@angular-devkit/build-angular:application",
          "options": {
            "allowedCommonJsDependencies": ["clipboard"],
            "outputPath": "dist/gestructureerde-mededeling",
            "index": "src/index.html",
            "browser": "src/main.ts",
            "tsConfig": "tsconfig.app.json",
            "inlineStyleLanguage": "scss",
            "assets": ["./src/favicon.ico", "./src/assets", "./src/_redirects"],
            "styles": [
              "normalize.css/normalize.css",
              "@angular/material/prebuilt-themes/azure-blue.css",
              "./src/styles.scss"
            ],
            "scripts": []
          },
          "configurations": {
            "production": {
              "budgets": [
                {
                  "type": "initial",
                  "maximumWarning": "500kB",
                  "maximumError": "1MB"
                },
                {
                  "type": "anyComponentStyle",
                  "maximumWarning": "2kB",
                  "maximumError": "4kB"
                }
              ],
              "outputHashing": "all"
            },
            "development": {
              "optimization": false,
              "extractLicenses": false,
              "sourceMap": true
            }
          },
          "defaultConfiguration": "production"
        },
        "serve": {
          "builder": "@angular-devkit/build-angular:dev-server",
          "configurations": {
            "production": {
              "buildTarget": "gestructureerde-mededeling:build:production"
            },
            "development": {
              "buildTarget": "gestructureerde-mededeling:build:development"
            }
          },
          "defaultConfiguration": "development"
        },
        "extract-i18n": {
          "builder": "@angular-devkit/build-angular:extract-i18n"
        },
        "test": {
          "builder": "@angular-devkit/build-angular:jest",
          "options": {
            "tsConfig": "tsconfig.spec.json"
          }
        }
      }
    }
  }
}
