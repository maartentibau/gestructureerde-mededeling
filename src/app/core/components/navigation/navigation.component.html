<mat-toolbar>
  <mat-toolbar-row>
    <h1>
      <a mat-button routerLink="/" title="Gestructureerde Mededeling" aria-label="Ga terug naar de homepage">{{
        appTitle$ | async
      }}</a>
    </h1>
    <span class="flex-spacer"></span>
    <nav>
      @if (isSmallScreen$ | async) {
        @for (item of navigation$ | async; track item; let i = $index) {
          <a
            [attr.data-navigation-id]="i"
            mat-icon-button
            routerLinkActive="mat-mdc-raised-button"
            [routerLink]="item.path"
            [title]="item.title"
            ><fa-icon [icon]="item.icon"></fa-icon
          ></a>
        }
      } @else {
        @for (item of navigation$ | async; track item; let i = $index) {
          <a [attr.data-navigation-id]="i" mat-button routerLinkActive="mat-mdc-raised-button" [routerLink]="item.path"
            ><fa-icon [icon]="item.icon"></fa-icon>{{ item.label }}</a
          >
        }
      }
    </nav>
  </mat-toolbar-row>
</mat-toolbar>
