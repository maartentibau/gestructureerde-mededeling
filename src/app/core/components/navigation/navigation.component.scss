mat-toolbar {
  background-color: #0c652b;
}

h1 {
  color: #fff;
}

nav {
  color: #fff;

  a {
    text-transform: uppercase;
    margin: 0 0.5em;

    &.mdc-button > ::ng-deep span.mdc-button__label {
      display: flex;
      justify-content: center;
      align-items: center;

      & > fa-icon {
        margin-right: 8px;
      }
    }

    &:hover:not(.mat-mdc-raised-button):not(.active) {
      background-color: #033313;
    }

    &.mat-mdc-raised-button {
      color: #216a38;

      &:hover {
        background-color: white;
      }
    }

    fa-icon {
      font-size: 1rem;
    }
  }
}
