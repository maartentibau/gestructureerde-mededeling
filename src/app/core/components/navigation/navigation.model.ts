import { IconProp } from '@fortawesome/fontawesome-svg-core';

export enum NavigationLabel {
  Long = 'long',
  Short = 'short',
}

export type NavigationEntity = {
  path: string[];
  label: { [NavigationLabel.Long]: string; [NavigationLabel.Short]: string };
  icon: IconProp;
  title: string;
};

export type MainNavigation = {
  path: string[];
  label: string;
  icon: IconProp;
  title: string;
}[];
