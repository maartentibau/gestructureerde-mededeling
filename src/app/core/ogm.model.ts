export interface OgmData {
  number: string;
  numberFormat: string;
}
