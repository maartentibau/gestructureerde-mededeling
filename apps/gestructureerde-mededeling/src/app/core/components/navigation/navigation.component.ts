import { ChangeDetectionStrategy, Component } from '@angular/core';
import { NavigationEntity, NavigationLabel } from './navigation.model';
import { Observable } from 'rxjs';
import { map, tap } from 'rxjs/operators';
import { Breakpoints } from '@angular/cdk/layout';
import { ScreenService } from '../../services/screen.service';
import { FaIconLibrary } from '@fortawesome/angular-fontawesome';
import { faCog, faShieldCheck, faStar } from '@fortawesome/pro-solid-svg-icons';

export const APP_NAVIGATION: NavigationEntity[] = [
  { path: ['/create'], label: { long: 'Zelf mededeling maken', short: 'Zelf maken' }, icon: ['fas', 'star'] },
  { path: ['/generate'], label: { long: 'Genereer mededeling', short: 'Genereren' }, icon: ['fas', 'cog'] },
  {
    path: ['/validate'],
    label: { long: 'Controleer mededeling', short: 'Controleren' },
    icon: ['fas', 'shield-check'],
  },
];

@Component({
  selector: 'ogm-navigation',
  templateUrl: './navigation.component.html',
  styleUrls: ['./navigation.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class NavigationComponent {
  readonly navigation$: Observable<Omit<NavigationEntity, 'label'>[]>;
  readonly isSmallScreen$: Observable<boolean>;
  readonly appTitle$: Observable<string>;

  constructor(private screenService: ScreenService, private faIconLibrary: FaIconLibrary) {
    this.isSmallScreen$ = this.screenService
      .observerBreakpoints()
      .pipe(
        map((breakpoints: { [key: string]: boolean }) =>
          Boolean(breakpoints[Breakpoints.XSmall] || breakpoints[Breakpoints.Small])
        )
      );

    this.navigation$ = this.screenService.observerBreakpoints().pipe(
      map((breakpoints: { [key: string]: boolean }) => {
        const labelSize: NavigationLabel = Boolean(
          breakpoints[Breakpoints.XSmall] || breakpoints[Breakpoints.Small] || breakpoints[Breakpoints.Medium]
        )
          ? NavigationLabel.Short
          : NavigationLabel.Long;

        return APP_NAVIGATION.map((navigation: NavigationEntity) => ({
          ...navigation,
          label: navigation.label[labelSize],
        }));
      })
    );

    this.appTitle$ = this.screenService.observerBreakpoints().pipe(
      tap(console.log),
      map((breakpoints: { [key: string]: boolean }) =>
        breakpoints[Breakpoints.XSmall] ? 'OGM' : 'Gestructureerde Mededeling'
      )
    );

    this.faIconLibrary.addIcons(faShieldCheck, faCog, faStar);
  }
}
