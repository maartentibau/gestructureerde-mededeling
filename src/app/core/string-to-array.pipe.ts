import { Pipe, PipeTransform } from '@angular/core';

@Pipe({
  standalone: true,
  name: 'stringToArray',
})
export class StringToArrayPipe implements PipeTransform {
  transform(value: number | string | null | undefined): string[] {
    if ((value && typeof value === 'string') || typeof value === 'number') {
      return Array.from(String(value));
    }

    return [];
  }
}
