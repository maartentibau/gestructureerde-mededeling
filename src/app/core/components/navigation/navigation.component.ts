import { Breakpoints } from '@angular/cdk/layout';
import { AsyncPipe } from '@angular/common';
import { ChangeDetectionStrategy, Component } from '@angular/core';
import { MatButtonModule } from '@angular/material/button';
import { MatToolbarModule } from '@angular/material/toolbar';
import { RouterLink, RouterLinkActive } from '@angular/router';
import { FaIconLibrary, FontAwesomeModule } from '@fortawesome/angular-fontawesome';
import { faCheckCircle, faCog, faStar } from '@fortawesome/free-solid-svg-icons';
import { Observable } from 'rxjs';
import { map } from 'rxjs/operators';
import { ScreenService } from '../../services/screen.service';
import { MainNavigation, NavigationEntity, NavigationLabel } from './navigation.model';

export const APP_NAVIGATION: NavigationEntity[] = [
  { path: ['/create'], label: { long: 'Zelf mededeling maken', short: 'Zelf maken' }, icon: ['fas', 'star'] },
  { path: ['/generate'], label: { long: 'Genereer mededeling', short: 'Genereren' }, icon: ['fas', 'cog'] },
  {
    path: ['/validate'],
    label: { long: 'Controleer mededeling', short: 'Controleren' },
    icon: ['fas', 'check-circle'],
  },
];

@Component({
  imports: [
    MatToolbarModule,
    MatButtonModule,
    AsyncPipe,
    FontAwesomeModule,
    MatButtonModule,
    RouterLink,
    RouterLinkActive,
  ],
  selector: 'ogm-navigation',
  templateUrl: './navigation.component.html',
  styleUrls: ['./navigation.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class NavigationComponent {
  readonly navigation$: Observable<MainNavigation>;
  readonly isSmallScreen$: Observable<boolean>;
  readonly appTitle$: Observable<string>;

  constructor(
    private screenService: ScreenService,
    private faIconLibrary: FaIconLibrary,
  ) {
    this.isSmallScreen$ = this.screenService
      .observerBreakpoints()
      .pipe(
        map((breakpoints: { [key: string]: boolean }) =>
          Boolean(breakpoints[Breakpoints.XSmall] || breakpoints[Breakpoints.Small]),
        ),
      );

    this.navigation$ = this.screenService.observerBreakpoints().pipe(
      map((breakpoints: { [key: string]: boolean }) => {
        const labelSize: NavigationLabel =
          breakpoints[Breakpoints.XSmall] || breakpoints[Breakpoints.Small] || breakpoints[Breakpoints.Medium]
            ? NavigationLabel.Short
            : NavigationLabel.Long;

        return APP_NAVIGATION.map((navigation: NavigationEntity) => ({
          ...navigation,
          label: navigation.label[labelSize],
        }));
      }),
    );

    this.appTitle$ = this.screenService
      .observerBreakpoints()
      .pipe(
        map((breakpoints: { [key: string]: boolean }) =>
          breakpoints[Breakpoints.XSmall] ? 'OGM' : 'Gestructureerde Mededeling - OGM',
        ),
      );

    this.faIconLibrary.addIcons(faCheckCircle, faCog, faStar);
  }
}
