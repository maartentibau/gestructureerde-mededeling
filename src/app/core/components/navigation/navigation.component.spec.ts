import { Breakpoints } from '@angular/cdk/layout';
import { AsyncPipe, JsonPipe } from '@angular/common';
import { Component, input } from '@angular/core';
import { ComponentFixture, TestBed } from '@angular/core/testing';
import { MatButtonModule } from '@angular/material/button';
import { MatToolbarModule } from '@angular/material/toolbar';
import { provideRouter, RouterLink, RouterLinkActive } from '@angular/router';
import { FaIconLibrary } from '@fortawesome/angular-fontawesome';
import { render } from '@testing-library/angular';
import { TestScheduler } from 'rxjs/testing';
import { ScreenService } from '../../services/screen.service';
import { APP_NAVIGATION, NavigationComponent } from './navigation.component';
import { NavigationEntity, NavigationLabel } from './navigation.model';

@Component({
  imports: [JsonPipe],
  selector: 'fa-icon',
  template: ` <div>{{ icon | json }}</div> `,
})
class MockFaIconComponent {
  icon = input<string | string[] | undefined>();
}

describe('NavigationComponent', () => {
  let component: NavigationComponent;
  let fixture: ComponentFixture<NavigationComponent>;

  let screenService: ScreenService;
  let testScheduler: TestScheduler;

  beforeEach(async () => {
    ({ component, fixture, screenService } = await setup());

    testScheduler = new TestScheduler((actual, expected) => {
      expect(actual).toEqual(expected);
    });
  });

  it('should create', () => {
    // act
    fixture.detectChanges();

    // assert
    expect(component).toBeTruthy();
    expect(fixture).toMatchSnapshot();
  });

  describe('isSmallScreen$', () => {
    [
      { breakpoint: Breakpoints.XSmall, smallScreen: true },
      { breakpoint: Breakpoints.Small, smallScreen: true },
      { breakpoint: Breakpoints.Medium, smallScreen: false },
      { breakpoint: Breakpoints.Large, smallScreen: false },
      { breakpoint: Breakpoints.XLarge, smallScreen: false },
    ].forEach(({ breakpoint, smallScreen }) => {
      it(`should emit ${smallScreen} when ${breakpoint}`, () => {
        testScheduler.run(({ cold, expectObservable }) => {
          // prepare
          const breakPoints$ = cold('a', { a: { [breakpoint]: true } });
          jest.spyOn(screenService, 'observerBreakpoints').mockReturnValue(breakPoints$);

          // Because the component uses the screenService.observerBreakpoints stream in it's constructor
          // we're ensuring the mocks are set up before creating the component.
          // act
          fixture = TestBed.createComponent(NavigationComponent);
          component = fixture.componentInstance;

          // check
          expectObservable(component.isSmallScreen$).toBe('a', { a: smallScreen });
        });
      });
    });
  });

  describe('navigation$', () => {
    [
      { breakpoint: Breakpoints.XSmall, labelSize: NavigationLabel.Short },
      { breakpoint: Breakpoints.Small, labelSize: NavigationLabel.Short },
      { breakpoint: Breakpoints.Medium, labelSize: NavigationLabel.Short },
      { breakpoint: Breakpoints.Large, labelSize: NavigationLabel.Long },
      { breakpoint: Breakpoints.XLarge, labelSize: NavigationLabel.Long },
    ].forEach(({ breakpoint, labelSize }) => {
      it(`should use label ${labelSize} when ${breakpoint}`, () => {
        testScheduler.run(({ cold, expectObservable }) => {
          // prepare
          const expectedNavigation = APP_NAVIGATION.map((navigation: NavigationEntity) => ({
            ...navigation,
            label: navigation.label[labelSize],
          }));

          const breakPoints$ = cold('a', { a: { [breakpoint]: true } });
          jest.spyOn(screenService, 'observerBreakpoints').mockReturnValue(breakPoints$);

          // Because the component uses the screenService.observerBreakpoints stream in it's constructor
          // we're ensuring the mocks are set up before creating the component.
          // act
          fixture = TestBed.createComponent(NavigationComponent);
          component = fixture.componentInstance;

          // check
          expectObservable(component.navigation$).toBe('a', { a: expectedNavigation });
        });
      });
    });
  });

  describe('appTitle$', () => {
    it.each`
      breakpoint            | title
      ${Breakpoints.XSmall} | ${'OGM'}
      ${Breakpoints.Small}  | ${'Gestructureerde Mededeling - OGM'}
      ${Breakpoints.Medium} | ${'Gestructureerde Mededeling - OGM'}
      ${Breakpoints.Large}  | ${'Gestructureerde Mededeling - OGM'}
      ${Breakpoints.XLarge} | ${'Gestructureerde Mededeling - OGM'}
    `(`should have $title with given $breakpoint`, ({ breakpoint, title }) => {
      testScheduler.run(({ cold, expectObservable }) => {
        // prepare
        const breakPoints$ = cold('a', { a: { [breakpoint]: true } });
        jest.spyOn(screenService, 'observerBreakpoints').mockReturnValue(breakPoints$);

        // Because the component uses the screenService.observerBreakpoints stream in it's constructor
        // we're ensuring the mocks are set up before creating the component.
        // act
        fixture = TestBed.createComponent(NavigationComponent);
        component = fixture.componentInstance;

        // check
        expectObservable(component.appTitle$).toBe('a', { a: title });
      });
    });
  });
});

const setup = async () => {
  const renderResult = await render(NavigationComponent, {
    componentImports: [MatToolbarModule, AsyncPipe, MatButtonModule, MockFaIconComponent, RouterLink, RouterLinkActive],
    providers: [
      { provide: ScreenService, useValue: { observerBreakpoints: () => ({ pipe: jest.fn() }) } },
      { provide: FaIconLibrary, useValue: { addIcons: jest.fn() } },
      provideRouter([]),
    ],
  });

  const screenService = TestBed.inject(ScreenService);

  return {
    ...renderResult,
    component: renderResult.fixture.componentInstance,
    screenService,
  };
};
