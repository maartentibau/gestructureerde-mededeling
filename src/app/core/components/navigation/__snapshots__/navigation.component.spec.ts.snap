// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavigationComponent should create 1`] = `
<ogm-navigation
  appTitle$="undefined"
  faIconLibrary={[Function Object]}
  isSmallScreen$="undefined"
  navigation$="undefined"
  screenService={[Function Object]}
>
  <mat-toolbar
    class="mat-toolbar mat-toolbar-multiple-rows"
  >
    <mat-toolbar-row
      class="mat-toolbar-row"
    >
      <h1>
        <a
          class="mdc-button mat-mdc-button _mat-animation-noopable mat-unthemed mat-mdc-button-base"
          href="/"
          mat-button=""
          mat-ripple-loader-class-name="mat-mdc-button-ripple"
          mat-ripple-loader-uninitialized=""
          routerlink="/"
        >
          <span
            class="mat-mdc-button-persistent-ripple mdc-button__ripple"
          />
          <span
            class="mdc-button__label"
          >
          </span>
          <span
            class="mat-focus-indicator"
          />
          <span
            class="mat-mdc-button-touch-target"
          />
        </a>
      </h1>
      <span
        class="flex-spacer"
      />
      <nav>
      </nav>
    </mat-toolbar-row>
  </mat-toolbar>
</ogm-navigation>
`;
